# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/ObjectGroup.cpp
# Commit: e15f1a6f5a38
# Full Hash: e15f1a6f5a38b276216c77f2827945b6f5ec4ea9
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-06 09:33:16
# Regressor Bug: 1689413
# File Overlap Count: 1
# Description:
#   Bug 1689413 part 9 - Move proto from ObjectGroup to BaseShape. r=jonco
#   
#   The ObjectGroup still stores the proto as well but we don't really use it anymore.
#   This is simpler than removing the group's proto now because that would result in
#   "empty" groups and the group table then doesn't make sense anymore.
# ==============================================================================

diff -r 95fa0ec36f7a -r e15f1a6f5a38 js/src/vm/ObjectGroup.cpp
--- a/js/src/vm/ObjectGroup.cpp	Fri Mar 05 19:10:08 2021 +0000
+++ b/js/src/vm/ObjectGroup.cpp	Fri Mar 05 19:10:08 2021 +0000
@@ -58,7 +58,7 @@
   MOZ_ASSERT_IF(proto.isObject(), !IsWindow(proto.toObject()));
 }
 
-void ObjectGroup::setProtoUnchecked(TaggedProto proto) {
+void ObjectGroup::setProtoUncheckedDeprecated(TaggedProto proto) {
   proto_ = proto;
   MOZ_ASSERT_IF(proto_.isObject() && proto_.toObject()->is<NativeObject>(),
                 proto_.toObject()->isDelegate());
@@ -96,7 +96,8 @@
   }
 
   global->setGroupRaw(group);
-  return true;
+
+  return JSObject::setProtoUnchecked(cx, global, proto);
 }
 
 /////////////////////////////////////////////////////////////////////
@@ -120,7 +121,7 @@
     explicit Lookup(TaggedProto proto) : proto(proto) {}
 
     explicit Lookup(const NewEntry& entry)
-        : proto(entry.group.unbarrieredGet()->proto()) {}
+        : proto(entry.group.unbarrieredGet()->protoDeprecated()) {}
   };
 
   bool needsSweep() { return IsAboutToBeFinalized(&group); }
@@ -148,7 +149,7 @@
 
   static inline bool match(const ObjectGroupRealm::NewEntry& key,
                            const Lookup& lookup) {
-    TaggedProto proto = key.group.unbarrieredGet()->proto();
+    TaggedProto proto = key.group.unbarrieredGet()->protoDeprecated();
     return MovableCellHasher<TaggedProto>::match(proto, lookup.proto);
   }
 };
@@ -171,7 +172,7 @@
 
 MOZ_ALWAYS_INLINE ObjectGroup* ObjectGroupRealm::DefaultNewGroupCache::lookup(
     const JSClass* clasp, TaggedProto proto) {
-  if (group_ && group_->proto() == proto) {
+  if (group_ && group_->protoDeprecated() == proto) {
     return group_;
   }
   return nullptr;
@@ -212,7 +213,7 @@
       table->lookupForAdd(ObjectGroupRealm::NewEntry::Lookup(proto));
   if (p) {
     ObjectGroup* group = p->group;
-    MOZ_ASSERT(group->proto() == proto);
+    MOZ_ASSERT(group->protoDeprecated() == proto);
     groups.defaultNewGroupCache.put(group);
     return group;
   }
@@ -296,12 +297,12 @@
         group = Forwarded(group);
         entry.group.set(group);
       }
-      TaggedProto proto = group->proto();
+      TaggedProto proto = group->protoDeprecated();
       if (proto.isObject() && IsForwarded(proto.toObject())) {
         proto = TaggedProto(Forwarded(proto.toObject()));
         // Update the group's proto here so that we are able to lookup
         // entries in this table before all object pointers are updated.
-        group->proto() = proto;
+        group->protoDeprecated() = proto;
       }
     }
   }
@@ -321,7 +322,7 @@
   for (auto r = table->all(); !r.empty(); r.popFront()) {
     NewEntry entry = r.front();
     CheckGCThingAfterMovingGC(entry.group.unbarrieredGet());
-    TaggedProto proto = entry.group.unbarrieredGet()->proto();
+    TaggedProto proto = entry.group.unbarrieredGet()->protoDeprecated();
     if (proto.isObject()) {
       CheckGCThingAfterMovingGC(proto.toObject());
     }